src/test/scala/firrtlTests/RegisterUpdateSpec.scala

```diff
@@ -0,0 +1,66 @@
+// See LICENSE for license details.
+
+package firrtlTests
+
+import firrtl._
+import firrtl.testutils._
+import FirrtlCheckers._
+import scala.util.matching.Regex
+
+class RegisterUpdateSpec extends FirrtlFlatSpec {
+  def compile(input: String): CircuitState =
+    (new VerilogCompiler).compileAndEmit(CircuitState(parse(input), ChirrtlForm), List.empty)
+  def compileBody(body: String) = {
+    val str = """
+      |circuit Test :
+      |  module Test :
+      |""".stripMargin + body.split("\n").mkString("    ", "\n    ", "")
+    compile(str)
+  }
+
+  "Register update logic" should "not duplicate common subtrees" in {
+    val result = compileBody(s"""
+      |input clock : Clock
+      |output io : { flip in : UInt<8>, flip a : UInt<1>, flip b : UInt<1>, flip c : UInt<1>, out : UInt<8>}
+      |reg r : UInt<8>, clock
+      |when io.a :
+      |  r <= io.in
+      |when io.b :
+      |  when io.c :
+      |    r <= UInt(2)
+      |io.out <= r""".stripMargin
+    )
+    val verilog = result.getEmittedCircuit.value
+    result  shouldNot containLine ("r <= io_in;")
+    verilog shouldNot include     ("if (io_a) begin")
+    result  should    containLine ("r <= _GEN_0;")
+  }
+
+  it should "not let duplicate subtrees on one register affect another" in {
```

Comment:
Do you think you could add a test that contains the expected firrtl after running the `FlattenRegUpdate` pass?
This would help by serving as a visual example of what `FlattenRegUpdate` does.

Similar to this: https://github.com/freechipsproject/firrtl/blob/master/src/test/scala/firrtlTests/WiringTests.scala#L71

Reply:
I think I can, it's a little more annoying because the output of `FlattenRegUpdate` doesn't escape the VerilogEmitter, but I should be able to use a TransformManager to get the exactly output of it.

Reply:
Pushing tests in a sec, for posterity, I tried to use the nice tree matchers but it's a bit too verbose:
```scala
    fstate should containTree {
      case Connect(_,
             Reference("r2",_,_,_),
             Mux(Reference("io_b",_,_,_),
                 UIntLiteral(x, _),
                 Mux(Reference("io_a",_,_,_),
                     Reference("io_in",_,_,_),
                     Reference("r2",_,_,_),
                     _
                 ),
                 _
             )
            ) if x == 3 => true
    }
```
vs.
```scala
fstate should containLine ("""r2 <= mux(io_b, UInt<8>("h3"), mux(io_a, io_in, r2))""")
```

Reply:
Done.